test/raft/unit/test_snapshot.c

```diff
@@ -359,26 +399,95 @@ TEST(snapshot_leader, timeouts, set_up, tear_down, 0, NULL) {
 
 	PRE(sm_state(&leader.sm) == LS_PAGE_READ);
 	ut_rpc_sent(&leader.rpc);
+	munit_assert_int(ut_get_msg_sent().type, ==, RAFT_IO_INSTALL_SNAPSHOT_CP);
 	ut_rpc_to_expired(&leader.rpc);
 
 	PRE(sm_state(&leader.sm) == LS_PAGE_READ);
 	ut_rpc_sent(&leader.rpc);
+	munit_assert_int(ut_get_msg_sent().type, ==, RAFT_IO_INSTALL_SNAPSHOT_CP);
 	ut_leader_message_received(&leader, ut_page_result());
 
 	PRE(sm_state(&leader.sm) == LS_SNAP_DONE);
 	ut_rpc_sent(&leader.rpc);
+	munit_assert_int(ut_get_msg_sent().type, ==, RAFT_IO_INSTALL_SNAPSHOT);
 	ut_leader_message_received(&leader, ut_install_snapshot_result());
 
 	sm_fini(&leader.sm);
 	return MUNIT_OK;
 }
 
+struct test_fixture {
+	union {
+		struct leader leader;
+		struct follower follower;
+	};
+	/* true when union contains leader, false when it contains follower */
+	bool is_leader;
+
+	/* We only expect one message to be in-flight. */
+	struct raft_message last_msg_sent;
+	bool msg_valid;
+
+	pool_t pool;
+
+	/* TODO: should accomodate several background jobs in the future. Probably
+	 * by scheduling a barrier in the pool after all the works that toggles this
+	 * flag. */
+	work_op orig_work_cb;
+	bool work_done;
+};
+
+/* Not problematic because each test runs in a different process. */
+static struct test_fixture global_fixture;
+
+#define MAGIC_MAIN_THREAD 0xdef1
+
+static __thread int thread_identifier;
+
+static void *pool_set_up(MUNIT_UNUSED const MunitParameter params[],
+                   MUNIT_UNUSED void *user_data)
+{
+	/* Prevent hangs. */
+	alarm(2);
+
+	pool_init(&global_fixture.pool, uv_default_loop(), 4, POOL_QOS_PRIO_FAIR);
+	global_fixture.pool.flags |= POOL_FOR_UT;
+	thread_identifier = MAGIC_MAIN_THREAD;
+
+	struct fixture *f = munit_malloc(sizeof *f);
+	return f;
+}
+
+static void pool_tear_down(void *data)
+{
+    free(data);
+}
+
 static void progress(void) {
-	for (unsigned i = 0; i < 100; i++) {
+	for (unsigned i = 0; i < 20; i++) {
+		uv_run(uv_default_loop(), UV_RUN_NOWAIT);
+	}
+}
+
+static void wait_work(void) {
+	while (!global_fixture.work_done) {
 		uv_run(uv_default_loop(), UV_RUN_NOWAIT);
 	}
 }
 
+static void wait_msg_sent(void) {
+	while (!global_fixture.msg_valid) {
+		uv_run(uv_default_loop(), UV_RUN_NOWAIT);
+	}
+}
+
+/* Decorates the callback used when the pool work is done to set the test
+ * fixture flag to true, then calls the original callback.*/
+static void test_fixture_work_cb(pool_work_t *w) {
+	global_fixture.work_done = true;
```

Comment:
Seems a good time to invent an implementation of `is_main_thread()` macro or a function is much closer than we thought... When I looked an `wait_work()` it wasn't obvious that the spin-lock runs in the same thread with `test_fixture_work_cb()`.

Please put a `// PRE(is_main_thread())` to both functions.

Reply:
I will also add some comments to the `global_fixture` and these functions.